### PR TITLE
feature: Vue Native wallet connection

### PR DESCRIPTION
**Pull request by 1saf**
Fixes # .

Changes proposed in this pull request:
- 
- 
-

### CODE REVIEW

File: src/components/cards/TradeCard/TradePair.vue [hunk 1]
```diff
@@ -149,7 +149,7 @@
 </template>
 
 <script lang="ts">
-import { defineComponent, toRefs, computed, ref } from 'vue';
+import { defineComponent, toRefs, computed, ref, watch } from 'vue';
```
Comment (evgenyboxer) [inline src/components/cards/TradeCard/TradePair.vue:152]: doesn't seem to be used?

File: src/components/cards/TradeCard/TradePair.vue [hunk 2]
```diff
@@ -158,6 +158,7 @@ import { ETHER } from '@/constants/tokenlists';
 
 import TradePairToggle from '@/components/cards/TradeCard/TradePairToggle.vue';
 import SelectTokenModal from '@/components/modals/SelectTokenModal/SelectTokenModal.vue';
+import useTokenLists from '@/composables/useTokenLists';
```
Comment (evgenyboxer) [inline src/components/cards/TradeCard/TradePair.vue:161]: This one as well (not used)

File: src/services/web3/connectors/connector.ts [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    this.selectedAccount = selectedAccount || '';
+  }
+
+  // must return the provider
+  abstract connect(): Promise<ConnectorPayload>;
+
+  handleAccountsChanged = accounts => {
+    if (this.selectedAccount !== '') {
+      const account = accounts.find(
+        a => getAddress(a) === getAddress(this.selectedAccount)
```
Comment (evgenyboxer) [inline src/services/web3/connectors/connector.ts:28]:
```suggestion
        account => getAddress(account) === getAddress(this.selectedAccount)
```

File: src/services/web3/connectors/metamask/metamask.connector.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    if (provider) {
+      this.provider = provider;
+      this.active.value = true;
+
+      try {
+        const accounts = await provider.request({
+          method: 'eth_requestAccounts'
+        });
+
+        const chainId = await provider.request({ method: 'eth_chainId' });
```
Comment (evgenyboxer) [inline src/services/web3/connectors/metamask/metamask.connector.ts:16]: Seems like `accounts` and `chainId` can be combined to a `promise.all`

File: src/services/web3/useVueWeb3.ts [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+      enabled: canLoadProfile
+    })
+  );
+
+  // TODO seperate this out?
+  // load up a list of all the evm chains
+  const { data: evmChains, isLoading: isLoadingEvmChains } = useQuery(
+    'EVM_CHAINS',
+    async () =>
+      await (await axios.get('https://chainid.network/chains.json')).data
```
Comment (evgenyboxer) [inline src/services/web3/useVueWeb3.ts:62]: double `await` ^

File: src/services/web3/connectors/connector.ts [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+  account: Ref<string | null>;
+  chainId: Ref<number | null>;
+};
+export abstract class Connector {
+  provider: any = null;
+  account: Ref<string | null> = ref(null);
+  chainId: Ref<number | null> = ref(null);
+  active: Ref<boolean> = ref(false);
+  selectedAccount = '';
+  id = 'unknown';
```
Comment (evgenyboxer) [inline src/services/web3/connectors/connector.ts:16]: I think we should add a `signer` here as well so that it's easily available when you need it. (its essentially calling `getSigner()` on the provider)

File: src/App.vue [hunk 7]
```diff
@@ -1,17 +1,16 @@
 <template>
+  <div id="modal" />
```
Comment (evgenyboxer) [inline src/App.vue:2]: Is this a fix for the modal behaviour?

File: src/components/forms/pool_actions/WithdrawForm.vue [hunk 8]
```diff
@@ -289,6 +294,7 @@ export default defineComponent({
     });
 
     const propMaxUSD = computed(() => {
+      console.log('ropmax', data.propMax);
```
Comment (evgenyboxer) [inline src/components/forms/pool_actions/WithdrawForm.vue:297]: remove `console.log`^

File: src/components/forms/pool_actions/WithdrawForm.vue [hunk 9]
```diff
@@ -313,7 +319,11 @@ export default defineComponent({
     });
 
     const bptBalance = computed(() => {
-      return allTokens.value[props.pool.address].balance;
+      console.log(
+        'tokendict',
+        tokenDictionary.value[props.pool.address].balance
+      );
```
Comment (evgenyboxer) [inline src/components/forms/pool_actions/WithdrawForm.vue:325]: remove `console.log`^

File: src/composables/useAllowances.ts [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+  );
+
+  const isQueryEnabled = computed(() => account && tokens.value.length > 0);
+  const {
+    data: allowances,
+    isLoading,
+    isFetching,
+    refetch: refetchAllowances
+  } = useQuery(
+    reactive(['ALLOWANCES', { chainId, account, dstList, tokens }]),
```
Comment (evgenyboxer) [inline src/composables/useAllowances.ts:45]: can `ALLOWANCES` be part of `queryKeys`

**Comment by evgenyboxer** [comment]
I've played a bit with the wallet and tried to make some account switches / account lock / etc, and I've had some errors thrown when trying to do it.

Screen Shot 2021-06-23 at 12 16 28 pm
Screen Shot 2021-06-23 at 12 18 52 pm
Screen Shot 2021-06-23 at 12 19 01 pm
Screen Shot 2021-06-23 at 12 11 57 pm

File: src/services/web3/useVueWeb3.ts [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+
+  const canLoadProfile = computed(
+    () => account.value !== '' && chainId.value !== 0
+  );
+
+  const getProvider = () => new Web3Provider(provider.value as any);
+
+  // TODO separate this out?
+  const { isLoading: isLoadingProfile, data: profile } = useQuery(
+    ['WEB3_PROFILE', { account, chainId }],
```
Comment (evgenyboxer) [inline src/services/web3/useVueWeb3.ts:50]: Can we have this in queryKeys?

File: src/services/web3/useVueWeb3.ts [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+    () => getProfile(account.value, String(chainId.value)),
+    reactive({
+      enabled: canLoadProfile
+    })
+  );
+
+  // TODO seperate this out?
+  // load up a list of all the evm chains
+  const { data: evmChains, isLoading: isLoadingEvmChains } = useQuery(
+    'EVM_CHAINS',
```
Comment (evgenyboxer) [inline src/services/web3/useVueWeb3.ts:60]: This one as well

File: src/services/web3/connectors/metamask/metamask.connector.ts (hunk 4 above)
Comment (1saf) [reply src/services/web3/connectors/metamask/metamask.connector.ts:16]: I could do this, but I prefer the readability of having them separate

File: src/services/web3/connectors/connector.ts (hunk 6 above)
Comment (1saf) [reply src/services/web3/connectors/connector.ts:16]: I believe most provider consumers will extract the signer for you if you simply provide a signer capable provider? I could be wrong

File: src/App.vue (hunk 7 above)
Comment (1saf) [reply src/App.vue:2]: Yes

**Comment by 1saf** [comment]
> I've played a bit with the wallet and tried to make some account switches / account lock / etc, and I've had some errors thrown when trying to do it.
> 
> Screen Shot 2021-06-23 at 12 16 28 pm
> 
> Screen Shot 2021-06-23 at 12 18 52 pm
> 
> Screen Shot 2021-06-23 at 12 19 01 pm
> 
> Screen Shot 2021-06-23 at 12 11 57 pm
> 
> Other than those errors it seems to connect way faster!

Resolved this :)

File: src/services/web3/connectors/connector.ts (hunk 6 above)
Comment (evgenyboxer) [reply src/services/web3/connectors/connector.ts:16]: That's an awesome idea, I wish it worked this way - but it seems you have to grab the signer from the provider to do wallet stuff (see `sendTransaction` function in `web3.ts`)

**Review by johngrantuk — APPROVED** [review]
Look good. Mentioned Trade balance issue to Ansaf directly.

File: src/services/web3/useVueWeb3.ts [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+  const {
+    connectWallet,
+    account,
+    chainId,
+    disconnectWallet,
+    connector,
+    provider,
+    walletState
+  } = inject(Web3ProviderSymbol) as Web3Plugin;
+  const appConfig = new ConfigService();
```
Comment (garethfuller) [inline src/services/web3/useVueWeb3.ts:22]:
```suggestion
  const configService = new ConfigService();
```

File: src/services/web3/useVueWeb3.ts [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+
+  const userNetworkConfig = computed(() =>
+    appConfig.getNetworkConfig(String(chainId.value))
```
Comment (garethfuller) [inline src/services/web3/useVueWeb3.ts:25]:
```suggestion
    configService.getNetworkConfig(String(chainId.value))
```

File: src/services/web3/useVueWeb3.ts [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+  );
+  const appNetworkConfig = appConfig.network;
```
Comment (garethfuller) [inline src/services/web3/useVueWeb3.ts:27]:
```suggestion
  const appNetworkConfig = configService.network;
```

File: src/services/web3/useVueWeb3.ts [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+
+  const networkName = computed(() => {
+    if (!isLoadingEvmChains.value) {
+      const chain = evmChains.value.find(
+        chain => chain.networkId === userNetworkConfig.value.chainId
+      );
+      return chain?.name || 'Unknown Network';
+    }
+    return 'Loading...';
+  });
```
Comment (garethfuller) [inline src/services/web3/useVueWeb3.ts:76]: Why do we need this? The networkName should come from the configs we already have. https://github.com/balancer-labs/frontend-v2/blob/develop/src/lib/config/homestead.json#L4

File: src/services/web3/useVueWeb3.ts [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+  const getProvider = () => new Web3Provider(provider.value as any);
+
+  // TODO separate this out?
+  const { isLoading: isLoadingProfile, data: profile } = useQuery(
+    QUERY_KEYS.Account.Profile(account, userNetworkConfig),
+    () => getProfile(account.value, String(userNetworkConfig.value.chainId)),
+    reactive({
+      enabled: canLoadProfile
+    })
+  );
```
Comment (garethfuller) [inline src/services/web3/useVueWeb3.ts:59]: I'd quite like to see this separated out. In the format we have the other queries. I know it's a really simple one, but if you're not familiar with reactQuery it's actually quite hard to parse and understand what's happening.

Also, could you adjust `getProfile` to not fetch the 3box data. We basically only want to fetch the ens name now.

File: src/services/web3/web3.plugin.ts [hunk 18]
```diff
@@ -0,0 +1,148 @@
...
+import { Web3Provider } from '@ethersproject/providers';
+import { WalletLinkConnector } from './connectors/walletlink/walletlink.connector';
+
+export type Wallet = 'metamask' | 'walletconnect' | 'walletlink';
+export const SupportedWallets = [
+  'metamask',
+  'walletconnect',
+  'walletlink'
+] as Wallet[];
+export const WalletNameDictionary: Record<Wallet, string> = {
```
Comment (garethfuller) [inline src/services/web3/web3.plugin.ts:16]: What do you think about using 'map' terminology instead of dictionary throughout the app? Dictionary is such a long word! 😅 
```suggestion
export const WalletNameMap: Record<Wallet, string> = {
```

File: src/services/web3/web3.plugin.ts [hunk 19]
```diff
@@ -0,0 +1,148 @@
...
+type WalletState = 'connecting' | 'connected' | 'disconnected';
+type PluginState = {
+  connector: any;
+  walletState: WalletState;
+};
+
+export default {
+  install: async app => {
+    const alreadyConnectedAccount = ref(lsGet('connectedWallet', null));
+    const alreadyConnectedProvider = ref(lsGet('connectedProvider', null));
```
Comment (garethfuller) [inline src/services/web3/web3.plugin.ts:49]: Nice! So this allows us to not have to reload the connection all the time right?

File: src/services/web3/web3.plugin.ts [hunk 20]
```diff
@@ -0,0 +1,148 @@
...
+      disconnectWallet,
+      ...toRefs(pluginState),
+      account,
+      chainId,
+      provider
+    };
+
+    app.provide(Web3ProviderSymbol, payload);
+  }
+};
```
Comment (garethfuller) [inline src/services/web3/web3.plugin.ts:148]: I'm ok with the logic here instead of abstracting elsewhere actually.

**Review by garethfuller — CHANGES_REQUESTED** [review]
Nice work, looks great! I've left some redundant comments (from our call) in the review just for the record.

File: src/services/web3/components/WalletSelectModal.vue [hunk 21]
```diff
@@ -0,0 +1,28 @@
+<template>
```
[the remark this answers is not in the document]
Comment (1saf) [reply src/services/web3/components/WalletSelectModal.vue]: This has been resolved, thanks for raising. Components are now part of the bal component library (non-global)

File: src/components/forms/pool_actions/InvestForm.vue [hunk 22]
```diff
@@ -561,16 +570,23 @@ export default defineComponent({
...
+          onTxConfirmed: async (tx: TransactionData) => {
             emit('success', tx);
             data.amounts = [];
             data.loading = false;
+            await refetchBalances.value();
+            setPropMax();
+            if (hasZeroBalance.value) {
+              data.investType = FormTypes.custom;
+              resetSlider();
+            }
```
[the remark this answers is not in the document]
Comment (1saf) [reply src/components/forms/pool_actions/InvestForm.vue:589]: I'll re-test and make sure we do not need this.

File: src/services/web3/useVueWeb3.ts (hunk 16 above)
Comment (1saf) [reply src/services/web3/useVueWeb3.ts:76]: I originally had this because if the network is unsupported - it will not show anything as part of the navbar settings modal. I can change it to use from the config.

File: src/services/web3/useVueWeb3.ts (hunk 17 above)
Comment (1saf) [reply src/services/web3/useVueWeb3.ts:59]: No worries.

File: src/services/web3/web3.plugin.ts (hunk 18 above)
Comment (1saf) [reply src/services/web3/web3.plugin.ts:16]: Done deal

File: src/services/web3/web3.plugin.ts (hunk 19 above)
Comment (1saf) [reply src/services/web3/web3.plugin.ts:49]: That is correct, it's mainly for when a user comes back to the app, it will load the provider + account that was previously used.

File: src/services/web3/useVueWeb3.ts (hunk 16 above)
Comment (garethfuller) [reply src/services/web3/useVueWeb3.ts]: Yeah I don't think we need to add another network request for this. Especially if we go with a full-page modal for unsupported networks. We can also just use the term 'Unsupported network' in that modal rather than fetching the network's name.

File: src/services/web3/useVueWeb3.ts (hunk 17 above)
Comment (1saf) [reply src/services/web3/useVueWeb3.ts]: Leaving this here actually, conversation around too much abstraction. It is a pretty simply query with not much ancillary compute happening, so will leave here. We can revisit to separate it out if it becomes a big boy

File: src/services/web3/useVueWeb3.ts (hunk 16 above)
Comment (1saf) [reply src/services/web3/useVueWeb3.ts]: All fixed

File: src/services/web3/connectors/connector.ts (hunk 6 above)
Comment (1saf) [reply src/services/web3/connectors/connector.ts:15]: added in a signer which gets returned from the composable :)

**Review by garethfuller — APPROVED** [review]
Should we not turn this off now: https://github.com/balancer-labs/frontend-v2/blob/378d67d838e39353989626e9dad02a0b302c63dc/src/store/modules/app.ts#L44-L46

**Comment by 1saf** [comment]
> Should we not turn this off now:
> 
> https://github.com/balancer-labs/frontend-v2/blob/378d67d838e39353989626e9dad02a0b302c63dc/src/store/modules/app.ts#L44-L46

yeos
